### PR TITLE
[mlir] Expose type and attribute names in the MLIRContext and abstract type/attr classes

### PR DESCRIPTION
**Pull request by math-fehr**
This patch expose the type and attribute names in C++ as methods in the `AbstractType` and `AbstractAttribute` classes, and keep a map of names to `AbstractType` and `AbstractAttribute` in the `MLIRContext`. Type and attribute names should be unique.

It adds support in ODS to generate the `getName` methods in `AbstractType` and `AbstractAttribute`, through the use of two new variables, `typeName` and `attrName`. It also adds names to C++-defined type and attributes.

### CODE REVIEW

**Comment by math-fehr** [comment]
RFC here: https://discourse.llvm.org/t/rfc-exposing-type-and-attribute-names-in-c/74896

File: mlir/include/mlir/Dialect/DLTI/DLTI.h [hunk 1]
```diff
@@ -55,6 +55,10 @@ class DataLayoutEntryAttr
 
   /// Prints this attribute.
   void print(AsmPrinter &os) const;
+
+  static constexpr StringRef getAttrName() {
```
Comment (joker-eph) [inline mlir/include/mlir/Dialect/DLTI/DLTI.h:59]: Why not just `getName()` so this is the same API on Types and Attribute (which is always more friendly to generic code).

File: mlir/unittests/Interfaces/DataLayoutInterfacesTest.cpp [hunk 2]
```diff
@@ -131,6 +138,8 @@ struct TypeNoLayout : public Type::TypeBase<TypeNoLayout, Type, TypeStorage> {
 
   using Base::Base;
 
+  static constexpr StringRef getTypeName() { return "test.no_layout"; }
+
```
Comment (joker-eph) [inline mlir/unittests/Interfaces/DataLayoutInterfacesTest.cpp:142]: One thing with these accessors is that they will interfere with the auto-generated accessors for the parameters, what happens in this case?
Should ODS forbid a parameter named `name`?

File: mlir/include/mlir/IR/BuiltinAttributes.td [hunk 3]
```diff
@@ -64,6 +64,7 @@ def Builtin_AffineMapAttr : Builtin_Attr<"AffineMap", [
     AffineMap getAffineMap() const { return getValue(); }
   }];
   let skipDefaultBuilders = 1;
+  let attrName = "builtin.affine_map";
```
Comment (makslevental) [inline mlir/include/mlir/IR/BuiltinAttributes.td:67]: So I'm just curious - why can't the same reparameterization that's done for `Builtin_Type` (taking the mnemonic as an arg) be done for `Builtin_Attr`?

File: mlir/examples/toy/Ch7/include/toy/Dialect.h [hunk 4]
```diff
@@ -72,6 +72,9 @@ class StructType : public mlir::Type::TypeBase<StructType, mlir::Type,
 
   /// Returns the number of element type held by this struct.
   size_t getNumElementTypes() { return getElementTypes().size(); }
+
+  /// Returns the name of this struct type.
+  static constexpr StringRef getTypeName() { return "toy.struct"; }
```
Comment (kuhar) [inline mlir/examples/toy/Ch7/include/toy/Dialect.h:77]: Why not `StringLiteral`?

File: mlir/include/mlir/IR/BuiltinAttributes.td (hunk 3 above)
Comment (math-fehr) [reply mlir/include/mlir/IR/BuiltinAttributes.td]: Oh yes, that's an overlook on my side, I fixed it now to work like `Builtin_Type`!

File: mlir/examples/toy/Ch7/include/toy/Dialect.h (hunk 4 above)
Comment (math-fehr) [reply mlir/examples/toy/Ch7/include/toy/Dialect.h:77]: Oh yes, that makes more sense, I'll fix that!

File: mlir/lib/IR/MLIRContext.cpp [hunk 5]
```diff
@@ -212,6 +212,9 @@ class MLIRContextImpl {
   DenseMap<TypeID, AbstractType *> registeredTypes;
   StorageUniquer typeUniquer;
 
+  /// This is a mapping from type name to the abstract type describing it.
+  llvm::StringMap<AbstractType *> nameToType;
```
Comment (joker-eph) [inline mlir/lib/IR/MLIRContext.cpp:216]: Can we use `DenseMap<StringAttr, AbstractType *> nameToType;` (more efficient storage, and more efficient lookup)?

Also when is it useful? (please add to the comment: the "what" is less important than the "why").

Comment (math-fehr) [reply mlir/lib/IR/MLIRContext.cpp]: Added!

**Comment by math-fehr** [comment]
I moved `getTypeName` and `getAttrName` to just a `name` field, so that it will be the same for both attributes and types, and also would not clash with any `name` parameter that could be define in ODS.
Do you think that make sense, or do you think that this is too much different than operations (that have a `getName` static method). Or do you think this should also be changed in operations?

Also, I added the change in `MLIRContext` so that we store a `DensMap` of `StringAttr` rather than a `StringMap`.

Currently, I have a `Address boundary error` when `build/NATIVE/bin/mlir-linalg-ods-yaml-gen` is being used in the build. I'm trying to figure out if it's because of the `::name` change, or the `StringAttr`.

File: mlir/include/mlir/Dialect/DLTI/DLTI.h (hunk 1 above)
Comment (joker-eph) [reply mlir/include/mlir/Dialect/DLTI/DLTI.h]: Reasonnable I guess

**Comment by math-fehr** [comment]
So my understanding is that `DenseMap<StringAttr, AbstractType*>` was what was creating the problem, but I'm not sure if this is a mistake on my end or if this is not possible.

For it to work, we would need to create a `StringAttr` before loading the `builtin` dialect, is that possible? This is because when we are loading the `builtin` dialect, we need to create those `StringAttr` for the type names when we load each type/attribute.

So for now I removed it, please tell me if you see a way I could make this work!

File: mlir/unittests/IR/TypeAttrNamesTest.cpp [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+  // Check that the abstract Attribute is the same as the one used by the
+  // Attribute.
+  auto barAttr = BarAttr::get(&ctx);
+  EXPECT_TRUE(&barAttr.getAbstractAttribute() == &barAbstractAttr->get());
+
+  // Check that lookups of non-existing Attributes returns nullopt.
+  // Even if an attribute with the same name exists.
+  EXPECT_FALSE(AbstractAttribute::lookup("fake.foo", &ctx).has_value());
+}
```
Comment (Mogball) [inline mlir/unittests/IR/TypeAttrNamesTest.cpp:90]: newline

**Review by Mogball — APPROVED** [review]

**Comment by math-fehr** [comment]
@joker-eph @kuhar, does this PR sounds good to you now?
I'll wait for your approval to merge it, but otherwise I think enough time passed that anyone interested commented?

File: mlir/lib/IR/MLIRContext.cpp (hunk 5 above)
Comment (joker-eph) [reply mlir/lib/IR/MLIRContext.cpp:216]: did you?

**Comment by joker-eph** [comment]
> So my understanding is that `DenseMap<StringAttr, AbstractType*>` was what was creating the problem, but I'm not sure if this is a mistake on my end or if this is not possible.
> 
> For it to work, we would need to create a `StringAttr` before loading the `builtin` dialect, is that possible? This is because when we are loading the `builtin` dialect, we need to create those `StringAttr` for the type names when we load each type/attribute.

Oh great... We can't add the name for `StringAttr` before registering the storage is that it?

Pretty sad about these string duplication, but we can live with it. Please document this next to the StringMap.

> So for now I removed it, please tell me if you see a way I could make this work!

It's likely possible, but will require some deeper look. I may try this after you land it!

File: mlir/lib/IR/MLIRContext.cpp (hunk 5 above)
Comment (math-fehr) [reply mlir/lib/IR/MLIRContext.cpp:216]: Oh my bad, it got erased again in a following PR, I'll add it back now!
Sorry about that!

**Comment by math-fehr** [comment]
> > So my understanding is that `DenseMap<StringAttr, AbstractType*>` was what was creating the problem, but I'm not sure if this is a mistake on my end or if this is not possible.
> > For it to work, we would need to create a `StringAttr` before loading the `builtin` dialect, is that possible? This is because when we are loading the `builtin` dialect, we need to create those `StringAttr` for the type names when we load each type/attribute.
> 
> Oh great... We can't add the name for `StringAttr` before registering the storage is that it?
> 
> Pretty sad about these string duplication, but we can live with it. Please document this next to the StringMap.
> 
> > So for now I removed it, please tell me if you see a way I could make this work!
> 
> It's likely possible, but will require some deeper look. I may try this after you land it!

Another solution would be to use a `DenseMap` of `StringRef` instead, as all names are basically `StringLiteral` for C++ defined attributes, and `StringRef` for dynamically defined attributes (as they already store their names).
Do you think that makes more sense for this PR?

**Review by kuhar — APPROVED** [review]
LGTM

**Comment by joker-eph** [comment]
> Do you think that makes more sense for this PR?

We could if we're confident this is enough. The API should then take a "StringLiteral" (it's more specific than a StringRef) I believe.

**Comment by math-fehr** [comment]
> > Do you think that makes more sense for this PR?
> 
> We could if we're confident this is enough. The API should then take a "StringLiteral" (it's more specific than a StringRef) I believe.

I'm confident `StringRef` will work, since it works for both statically-defined and dynamically defined types and attributes.
For `StringLiteral`, it wouldn't work for dynamically defined attributes and types.

**Comment by joker-eph** [comment]
Well it's a low-level API, so let's document the lifetime constraint on the StringRef then!

**Review by joker-eph — APPROVED** [review]

**Comment by makslevental** [comment]
Congrats @math-fehr on landing a solid patch. I'll be using/abusing this soon for the Python bindings :)

**Pull request merged**